web_generator/lib/src/interop_gen/transform/transformer.dart

```diff
@@ -0,0 +1,214 @@
+import 'dart:js_interop';
+import '../../ast.dart';
+import '../../ast_extensions.dart';
+import '../../js/typescript.dart' as ts;
+import '../../js/typescript.types.dart';
+import '../namer.dart';
+import '../transform.dart';
+
+class Transformer {
+  /// A map of already resolved TS Nodes
+  final List<TSNode> nodes = [];
+
+  /// A map of declarations
+  final DeclarationMap declarationMap = DeclarationMap(<String, Node>{});
+
+  /// The type checker for the given program
+  final ts.TSTypeChecker typeChecker;
+
+  /// A set of declarations to export
+  final Set<String> exportSet;
+
+  /// namer, for giving elements unique names
+  final UniqueNamer namer;
+
+  final ProgramDeclarationMap programMap;
+
+  Transformer(this.programMap, this.typeChecker,
+      {Iterable<String> exportSet = const []})
+      : exportSet = exportSet.toSet(),
+        namer = UniqueNamer();
+
+  void transform(TSNode node) {
+    if (nodes.contains(node)) return;
+
+    switch (node.kind) {
+      case TSSyntaxKind.VariableStatement:
+        final decs = _transformVariable(node as TSVariableStatement);
+        declarationMap.addAll({for (final d in decs) d.id: d});
+      default:
+        final Declaration decl = switch (node.kind) {
+          _ => throw Exception('Unsupported Declaration Kind: ${node.kind}')
+        };
+        // ignore: dead_code This line will not be dead in future decl additions
+        declarationMap[decl.id] = decl;
+    }
+
+    nodes.add(node);
+  }
+
+  DeclarationMap filter([bool Function(Node)? saveIf]) {
+    // Set<String> dependencies = {};
+    final filteredDeclarations = DeclarationMap({});
+
+    // filter out for export declarations
+    declarationMap.forEach((k, v) {
+      // get decls exported by name
+      if (exportSet.contains(v.name)) {
+        filteredDeclarations[k] = v;
+      }
+
+      // get decls with `export` keyword
+      switch (v) {
+        case final ExportableDeclaration e:
+          if (e.exported) {
+            filteredDeclarations[(e as Node).id] = e as Node;
+          }
+          break;
+        case final PrimitiveType _:
+          // primitive types are generated by default
+          break;
+        case Type():
+          // TODO: Handle this case.
+          throw UnimplementedError();
+        case Declaration():
+          // TODO: Handle this case.
+          throw UnimplementedError();
+      }
+    });
+
+    filteredDeclarations.removeWhere((k, v) => !(saveIf?.call(v) ?? true));
+
+    // then filter for dependencies
+    final otherDecls = filteredDeclarations.entries
+        .map((e) => _filterForDecl(e.value))
+        .reduce((value, element) => value..addAll(element));
+
+    return filteredDeclarations..addAll(otherDecls);
+  }
+
+  /// Given an already filtered declaration [decl],
+  /// filter out dependencies of [decl] and return them as a declaration map
+  DeclarationMap _filterForDecl([Node? decl]) {
```

Comment:
Maybe a slightly more descriptive name like `_getDependenciesOfDecl`. I'd also update the doc to reflect that this is recursive.